Sources/Features/Network/MultipartRequest.swift

```diff
@@ -48,91 +48,118 @@ extension MultipartRequest {
         var multipartRequest = urlRequest
         let boundary = generateBoundaryString()
         multipartRequest.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
-        
-        // temporary file to stream data
-        let tempURL = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent(UUID().uuidString)
-        
+
+        // output stream
+        // TODO: bind input & output streams
+        let outputStream = OutputStream.toMemory()
+        outputStream.open()
+        defer { outputStream.close() }
+
+        let writeLock = DispatchQueue(label: "backtrace.multipartRequest.writeLock")
+
         do {
-            let fileCreated = FileManager.default.createFile(atPath: tempURL.path, contents: nil, attributes: nil)
-            if !fileCreated {
-                throw HttpError.fileCreationFailed(tempURL)
-            }
-            
-            let fileHandle = try FileHandle(forWritingTo: tempURL)
-            defer {
-                if #available(iOS 13.0, tvOS 13.0, macOS 11.0, *) {
-                    try? fileHandle.close()
-                } else {
-                    fileHandle.closeFile()
-                }
-            }
-            
             // attributes
             var attributesString = ""
             for attribute in report.attributes {
                 attributesString += "--\(boundary)\r\n"
                 attributesString += "Content-Disposition: form-data; name=\"\(attribute.key)\"\r\n\r\n"
                 attributesString += "\(attribute.value)\r\n"
             }
-            try writeToFile(fileHandle, attributesString)
-            
+            try writeToStream(outputStream, attributesString, writeLock: writeLock)
+
             // report
             var reportString = "--\(boundary)\r\n"
             reportString += "Content-Disposition: form-data; name=\"upload_file\"; filename=\"upload_file\"\r\n"
             reportString += "Content-Type: application/octet-stream\r\n\r\n"
-            try writeToFile(fileHandle, reportString)
-            fileHandle.write(report.reportData)
-            try writeToFile(fileHandle, "\r\n")
-            
+            try writeToStream(outputStream, reportString, writeLock: writeLock)
+            try writeLock.sync {
+                let data = report.reportData
+                let bytesWritten = data.withUnsafeBytes {
```

Comment:
why do we need to execute this part of the code in the sync context? You won't submit two identical reports.